alibi_detect/cd/pytorch/spot_the_diff.py

```diff
@@ -0,0 +1,213 @@
+            for each feature with equal variance to that of reference data.
+        l1_reg
+            Strength of l1 regularisation to apply to the differences.
+        binarize_preds
+            Whether to test for discrepency on soft (e.g. probs/logits) model predictions directly
+            with a K-S test or binarise to 0-1 prediction errors and apply a binomial test.
+        train_size
+            Optional fraction (float between 0 and 1) of the dataset used to train the classifier.
+            The drift is detected on `1 - train_size`. Cannot be used in combination with `n_folds`.
+        n_folds
+            Optional number of stratified folds used for training. The model preds are then calculated
+            on all the out-of-fold instances. This allows to leverage all the reference and test data
+            for drift detection at the expense of longer computation. If both `train_size` and `n_folds`
+            are specified, `n_folds` is prioritized.
+        retrain_from_scratch
+            Whether the classifier should be retrained from scratch for each set of test data or whether
+            it should instead continue training from where it left off on the previous set.
+        seed
+            Optional random seed for fold selection.
+        optimizer
+            Optimizer used during training of the classifier.
+        learning_rate
+            Learning rate used by optimizer.
+        batch_size
+            Batch size used during training of the classifier.
+        preprocess_batch_fn
+            Optional batch preprocessing function. For example to convert a list of objects to a batch which can be
+            processed by the model.
+        epochs
+            Number of training epochs for the classifier for each (optional) fold.
+        verbose
+            Verbosity level during the training of the classifier. 0 is silent, 1 a progress bar.
+        train_kwargs
+            Optional additional kwargs when fitting the classifier.
+        device
+            Device type used. The default None tries to use the GPU and falls back on CPU if needed.
+            Can be specified by passing either 'cuda', 'gpu' or 'cpu'.
+        dataset
+            Dataset object used during training.
+        dataloader
+            Dataloader object used during training.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        if preprocess_fn is not None and preprocess_batch_fn is not None:
+            raise ValueError("SpotTheDiffDrift detector only supports preprocess_fn or preprocess_batch_fn, not both.")
+
+        if preprocess_fn is not None:
+            x_ref_proc = preprocess_fn(x_ref)
+        elif preprocess_batch_fn is not None:
+            x_ref_proc = predict_batch(
+                x_ref, lambda x: x, preprocess_fn=preprocess_batch_fn,
+                device=torch.device('cpu'), batch_size=batch_size
```

Comment:
We can infer the device already and then pass `device.type` to the `ClassifierDriftTorch` detector.

Reply:
But that's for applying the model whereas this is just for applying `lambda x: x`.

Reply:
Given the preds are instantly returned to the cpu it seemed silly potentially transferring them to a gpu to apply the identity.

Reply:
Ah yes, I forgot that the device comes only in use after the preprocessing in predict batch, my bad!